### PR TITLE
Add expansion rule to add cast to lambda expression bodies

### LINKED ISSUE #6306: Rename doesn't resolve conflicts by value of uncastable lambda
```C#
void X<T>(T t, Func<T, long> e) { X(new { a = 1 }, a => a.a); }
[Obsolete]
void Y<T>(T t, Func<T, int> e) { }
```

Rename `Y` to `X`.

<strike>
**Expected**:
```C#
X(new { a = (long)1 }, a => a.a);
```
</strike>

**Expected**:
```C#
X(new { a = 1 }, a => (long)a.a);
```

### LINKED ISSUE #6308: Rename doesn't resolve conflicts by inserting identity conversion
```C#
[Obsolete]
void M<T>(T t, Func<T,int, int> e) { }
int M<T>(T t, Func<T, long, long> e) => M(new { }, (_, a) => X(a));

long X(long a) => a;
int Y(int a) => a;
```

Rename `Y` to `X`.

**Expected**:
```C#
M(new { }, (_, a) => X((long)a));
```

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes #6306
Fixes #6308

This change also fixes two outstanding (though not yet reported) bugs in the expansion engine:

* Generic type arguments should *not* be addded to a name if they contain anonymous types
* A lambda parameter should not be expanded if its type is an anonymous type

Tagging @dotnet/roslyn-ide

### CODE REVIEW

**Comment by DustinCampbell** [comment]
Ping @dotnet/roslyn-ide

**Comment by Pilchie** [comment]
What about a couple of tests for those unreported bugs?

Otherwise :shipit:

**Comment by DustinCampbell** [comment]
test win/dbg/unit64 please

**Comment by DustinCampbell** [comment]
retest prtest/win/dbg/unit64 please

**Comment by DustinCampbell** [comment]
retest prtest/win/dbg/unit64 please

**Comment by DustinCampbell** [comment]
retest prtest/mac/dbg/unit32 please

**Comment by DustinCampbell** [comment]
Ping @roslyn-ide -- anyone else want to take a look?

File: src/EditorFeatures/Test2/Expansion/NameExpansionTests.vb [hunk 1]
```diff
@@ -148,6 +148,119 @@ namespace NS
...
+    <Project Language="C#" CommonReferences="true">
+        <Document>
+using System;
+class C
+{
+    static void Mumble&lt;T&gt;(T anonymousType, Action&lt;T, int, int&gt; lambda) { }
+
+    static void M()
+    {
+        Mumble(new { x = 42 }, {|Expand:(a, y) => a.x|});
```
Comment (zanqi) [inline src/EditorFeatures/Test2/Expansion/NameExpansionTests.vb]: Is the lambda expecting three parameters here to compile? It's a Action<T, int, int>, not Func<T, int, int>

**Pull request merged**